diff --git a/ruy/build_defs.oss.bzl b/ruy/build_defs.oss.bzl
index e405b41..1d7612b 100644
--- a/ruy/build_defs.oss.bzl
+++ b/ruy/build_defs.oss.bzl
@@ -11,5 +11,5 @@ def ruy_linkopts_thread_standard_library():
     # https://github.com/abseil/abseil-cpp/blob/1112609635037a32435de7aa70a9188dcb591458/absl/base/BUILD.bazel#L155
     return select({
         "@bazel_tools//src/conditions:windows": [],
-        "//conditions:default": ["-pthread"],
+        "//conditions:default": [],
     })